pyresample/bilinear/__init__.py

```diff
@@ -309,267 +653,322 @@ def _get_ts_irregular(pt_1, pt_2, pt_3, pt_4, out_y, out_x):
     return t__, s__
 
 
-def _get_ts_uprights_parallel(pt_1, pt_2, pt_3, pt_4, out_y, out_x):
-    """Get parameters for the case where uprights are parallel"""
+def _solve_quadratic(a__, b__, c__, min_val=0.0, max_val=1.0):
+    """Solve quadratic equation.
 
-    # Get parameters for the quadratic equation
-    a__, b__, c__ = _calc_abc(pt_1, pt_3, pt_2, pt_4, out_y, out_x)
+    Solve quadratic equation and return the valid roots from interval
+    [*min_val*, *max_val*].
 
-    # Get the valid roots from interval [0, 1]
-    s__ = _solve_quadratic(a__, b__, c__, min_val=0., max_val=1.)
+    """
+    discriminant = b__ * b__ - 4 * a__ * c__
 
-    # Calculate parameter t
-    t__ = _solve_another_fractional_distance(s__, pt_1[:, 1], pt_2[:, 1],
-                                             pt_3[:, 1], pt_4[:, 1], out_y)
+    # Solve the quadratic polynomial
+    x_1 = (-b__ + np.sqrt(discriminant)) / (2 * a__)
+    x_2 = (-b__ - np.sqrt(discriminant)) / (2 * a__)
 
-    return t__, s__
+    # Find valid solutions, ie. 0 <= t <= 1
+    x__ = np.where(
+        find_indices_outside_min_and_max(x_1, min_val, max_val),
+        x_2, x_1)
 
+    x__ = np.where(
+        find_indices_outside_min_and_max(x__, min_val, max_val),
+        np.nan, x__)
+
+    return x__
 
-def _get_ts_parallellogram(pt_1, pt_2, pt_3, out_y, out_x):
-    """Get parameters for the case where uprights are parallel"""
 
+def _calc_abc(pt_1, pt_2, pt_3, pt_4, out_y, out_x):
+    """Calculate coefficients for quadratic equation.
+
+    In this order of arguments used for _get_fractional_distances_irregular() and
+    _get_fractional_distances_uprights().  For _get_fractional_distances_uprights switch order of pt_2 and
+    pt_3.
+
+    """
     # Pairwise longitudal separations between reference points
     x_21 = pt_2[:, 0] - pt_1[:, 0]
     x_31 = pt_3[:, 0] - pt_1[:, 0]
+    x_42 = pt_4[:, 0] - pt_2[:, 0]
 
     # Pairwise latitudal separations between reference points
     y_21 = pt_2[:, 1] - pt_1[:, 1]
     y_31 = pt_3[:, 1] - pt_1[:, 1]
+    y_42 = pt_4[:, 1] - pt_2[:, 1]
 
-    t__ = (x_21 * (out_y - pt_1[:, 1]) - y_21 * (out_x - pt_1[:, 0])) / \
-          (x_21 * y_31 - y_21 * x_31)
-    with np.errstate(invalid='ignore'):
-        idxs = (t__ < 0.) | (t__ > 1.)
-    t__[idxs] = np.nan
-
-    s__ = (out_x - pt_1[:, 0] + x_31 * t__) / x_21
-
-    with np.errstate(invalid='ignore'):
-        idxs = (s__ < 0.) | (s__ > 1.)
-    s__[idxs] = np.nan
+    a__ = x_31 * y_42 - y_31 * x_42
+    b__ = out_y * (x_42 - x_31) - out_x * (y_42 - y_31) + \
+        x_31 * pt_2[:, 1] - y_31 * pt_2[:, 0] + \
+        y_42 * pt_1[:, 0] - x_42 * pt_1[:, 1]
+    c__ = out_y * x_21 - out_x * y_21 + pt_1[:, 0] * pt_2[:, 1] - \
+        pt_2[:, 0] * pt_1[:, 1]
```

Comment:
could `a__`, `b__`, `c__` be replaced with more explicit names? (genuine question)

Reply:
Couldn't think of a name for them. In the quadratic equation the symbols are `a`, `b` and `c`, so went with those (with three characters) when I implemented the original version couple of years ago.